lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -1736,10 +1738,19 @@ ParseResult FIRStmtParser::parseOptionalExpPostscript(Value &result) {
       continue;
     }
 
-    // Subindex: exp ::= exp '[' intLit ']' | exp '[' exp ']'
+    // Subindex: exp ::= exp '[' intLit ']' | exp '[' intLit ':' intLit ']' |
+    // exp '[' exp ']'
     if (consumeIf(FIRToken::l_square)) {
       if (getToken().isAny(FIRToken::integer, FIRToken::string)) {
-        if (parsePostFixIntSubscript(result))
+        // need one token lookahead here to see if this is a range subindex or
+        // not
+        auto backtrackState = getLexer().getCursor();
+        consumeToken();
+        bool range = getToken().is(FIRToken::colon);
+        backtrackState.restore(getLexer());
+        if (!range && parsePostFixIntSubscript(result))
+          return failure();
+        if (range && parsePostFixIntRangeSubscript(result))
```

Comment:
I'm not a fan of reparsing things, can we structure it more like this?  (maybe split it up into more functions, its getting a little long).

```c++
ParseResult FIRStmtParser::parsePostFixIntSubscript(Value &result) {
  auto loc = getToken().getLoc();
  int32_t indexNo;
  if (parseIntLit(indexNo, "expected index"))
    return failure();
  if (indexNo < 0)
    return emitError(loc, "invalid index specifier"), failure();

  // Make sure the index expression is valid and compute the result type for the
  // expression.
  // TODO: This should ideally be folded into a `tryCreate` method on the
  // builder (https://llvm.discourse.group/t/3504).

  if (result.getType().isa<IntType>()) {
    NamedAttrList bitAttrs;
    bitAttrs.append(getConstants().hiIdentifier,
                    builder.getI32IntegerAttr(indexNo));
    if (consumeIf(FIRToken::colon)) {
      if (parseIntLit(indexNo, "expected low index"))
        return failure();
    }
    bitAttrs.append(getConstants().loIdentifier,
                    builder.getI32IntegerAttr(indexNo));

    auto resultType = BitsPrimOp::inferReturnType({result}, bitAttrs, {});
    if (!resultType) {
      (void)BitsPrimOp::inferReturnType({result}, bitAttrs,
                                        translateLocation(loc));
      return failure();
    }
    result = builder.create<BitsPrimOp>(resultType, result, bitAttrs);
    return success();
  } else {
    NamedAttribute attrs = {getConstants().indexIdentifier,
                            builder.getI32IntegerAttr(indexNo)};
    auto resultType = SubindexOp::inferReturnType({result}, attrs, {});
    if (!resultType) {
      (void)SubindexOp::inferReturnType({result}, attrs,
                                        translateLocation(loc));
      return failure();
    }
    // Check if we already have created this Subindex op.
    auto &value = moduleContext.getCachedSubaccess(result, indexNo);
    if (value) {
      result = value;
      return success();
    }

    // Create the result operation, inserting at the location of the
    // declaration. This will cache the subindex operation for further uses.
    locationProcessor.setLoc(loc);
    OpBuilder::InsertionGuard guard(builder);
    builder.setInsertionPointAfterValue(result);

    auto op = builder.create<SubindexOp>(resultType, result, attrs);

    // Insert the newly created operation into the cache.
    value = op.getResult();
    result = value;
  }

  if (parseToken(FIRToken::r_square, "expected ']'"))
    return failure();
  return success();
}
```